### PR TITLE
improved queryindex performance when there is more than one filter. b…

### PR DESCRIPTION
**Pull request by OfirMos**
…ug fix: MOD-5338
On my M1 laptop the command:
`ts.queryindex min_purpose=(total,mt_total,success,mt_success) min_account=(26c66508,9184c342,e584b1d4,ab8299ca,a7ed404c,5fae980b,b4ce1b49)`
On approximately 2.5 million keys.
Before the change it took 70 seconds and after 2 seconds.

### CODE REVIEW

File: src/indexer.c [hunk 1]
```diff
@@ -240,248 +240,174 @@ int IsKeyIndexed(RedisModuleString *ts_key) {
...
     }
 
-    if (result == NULL) {
-        return RedisModule_CreateDict(ctx);
-    }
+    free(dicts);
 
     if (unlikely(isTrimming)) {
-        RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(result, "^", NULL, 0);
+        RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(res, "^", NULL, 0);
```
Comment (danni-m) [inline src/indexer.c:395]: Looks like we don't have tests to cover this, is it covered in other places?

File: src/indexer.c [hunk 2]
```diff
@@ -240,248 +240,174 @@ int IsKeyIndexed(RedisModuleString *ts_key) {
...
             }
         }
+        free(dicts);
 
-        // switch between the minimal predicate and the predicate in the first place
-        if (minIndex != 0) {
-            QueryPredicate temp = index_predicate[minIndex];
-            index_predicate[minIndex] = index_predicate[0];
-            index_predicate[0] = temp;
+        if (inclusion != found) {
```
[the remark this answers is not in the document]
Comment (OfirMos) [reply src/indexer.c:350]: @danni-m I think that || is better than adding another case, WDYT?

File: src/indexer.c (hunk 1 above)
Comment (OfirMos) [reply src/indexer.c:395]: @danni-m I think we this covered in the Automation tests.

**Comment by filipecosta90** [comment]
### Automated performance analysis summary

This comment was automatically generated given there is performance data available.

In summary:
- Detected a total of 6 stable tests between versions.
- Detected a total of 2 regressions bellow the regression water line 5.0.

You can check a comparison in detail via the [grafana link](https://benchmarksrediscom.grafana.net/d//2WMw61UGz/?var-branch=master&var-branch=perf_improve_queryindex)

### Comparison between master and perf_improve_queryindex.

Time Period from 30 days ago. (environment used: oss-standalone)

|                 Test Case                  |Baseline master (median obs. +- std.dev)|Comparison perf_improve_queryindex (median obs. +- std.dev)|% change (higher-better)|           Note           |
|--------------------------------------------|----------------------------------------|-----------------------------------------------------------|------------------------|--------------------------|
|ts_mget_100K-series-tsbs-devops             | 47 +- 8.2%  (4 datapoints)             | 38 +- nan%  (1 datapoints)                                |-18.0%                  |waterline=8.2%. REGRESSION|
|ts_mrevrange_as_mget_100K-series-tsbs-devops| 28 +- 9.6%  (4 datapoints)             | 21 +- nan%  (1 datapoints)                                |-23.3%                  |waterline=9.6%. REGRESSION|
|tsbs-scale100_cpu-max-all-1                 | 1565 +- 3.9%  (7 datapoints)           | 1619 +- nan%  (1 datapoints)                              |3.4%                    |potential IMPROVEMENT     |
|tsbs-scale100_cpu-max-all-8                 | 214 +- 4.0%  (7 datapoints)            | 218 +- nan%  (1 datapoints)                               |2.0%                    |-- no change --           |
|tsbs-scale100_double-groupby-1              | 104 +- 0.9%  (7 datapoints)            | 104 +- nan%  (1 datapoints)                               |0.5%                    |-- no change --           |
|tsbs-scale100_double-groupby-5              | 20 +- 3.5%  (7 datapoints)             | 21 +- nan%  (1 datapoints)                                |4.2%                    |potential IMPROVEMENT     |
|tsbs-scale100_double-groupby-all            | 10 +- 2.5%  (7 datapoints)             | 10 +- nan%  (1 datapoints)                                |2.3%                    |-- no change --           |
|tsbs-scale100_groupby-orderby-limit         | 1660 +- 1.0%  (7 datapoints)           | 1593 +- nan%  (1 datapoints)                              |-4.0%                   |potential REGRESSION      |

**Review by rafie — DISMISSED** [review]

**Review by danni-m — APPROVED** [review]

**Review by raffapen — APPROVED** [review]

**Pull request merged**